EmailValidator/Validation/MailboxCheckValidation.php

```diff
@@ -0,0 +1,240 @@
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function isValid($email, EmailLexer $emailLexer)
+    {
+        $mxHosts = $this->getMXHosts($email);
+
+        $isValid = false;
+        foreach ($mxHosts as $mxHost) {
+            if ( ($isValid = $this->checkMailbox($mxHost, $this->port, $this->timeout, $this->fromEmail, $email)) ) {
+                break;
+            }
+        }
+
+        if ( ! $isValid ) {
+            $this->error = new IllegalMailbox($this->lastResponseCode);
+        }
+
+        return $this->error === null;
+    }
+
+    /**
+     * @param string $email
+     *
+     * @return array
+     */
+    protected function getMXHosts($email)
+    {
+        $variant = defined('INTL_IDNA_VARIANT_UTS46') ? INTL_IDNA_VARIANT_UTS46 : INTL_IDNA_VARIANT_2003;
+
+        $hostname = $email;
+        if ( false !== ($lastAtPos = strrpos($email, '@')) ) {
+            $hostname = substr($email, $lastAtPos + 1);
+        }
+        $hostname = rtrim(idn_to_ascii($hostname, IDNA_DEFAULT, $variant), '.') . '.';
+
+        $mxHosts = [];
+        $result = getmxrr($hostname, $mxHosts);
+        if ( ! $result ) {
+            $this->warnings[NoDNSMXRecord::CODE] = new NoDNSMXRecord();
+        }
+
+        return $mxHosts;
+    }
+
+    /**
+     * @param string $hostname
+     * @param int $port
+     * @param int $timeout
+     * @param string $fromEmail
+     * @param string $toEmail
+     * @return bool
+     */
+    protected function checkMailbox($hostname, $port, $timeout, $fromEmail, $toEmail)
+    {
+        $socket = @fsockopen($hostname, $port, $errno, $errstr, $timeout);
+
+        if ( ! $socket ) {
+            $this->warnings[SocketWarning::CODE][] = new SocketWarning($hostname, $errno, $errstr);
+
+            return false;
+        }
+
+        if ( ! ($this->assertResponse($socket, 220) ) ) {
+            return false;
+        }
+
+        fwrite($socket, "EHLO {$hostname}" . self::END_OF_LINE);
+        if ( ! ($this->assertResponse($socket, 250) ) ) {
+            return false;
+        }
+
+        fwrite($socket, "MAIL FROM: <{$fromEmail}>" . self::END_OF_LINE);
+        if ( ! ($this->assertResponse($socket, 250) ) ) {
+            return false;
+        }
+
+        fwrite($socket, "RCPT TO: <{$toEmail}>" . self::END_OF_LINE);
+        if ( ! ($this->assertResponse($socket, 250) ) ) {
+            return false;
+        }
+
+        fwrite($socket, 'QUIT' . self::END_OF_LINE);
+
+        fclose($socket);
+
+        return true;
+    }
+
+    /**
+     * @param resource $socket
+     * @param int $expectedCode
+     *
+     * @return bool
+     */
+    private function assertResponse($socket, $expectedCode)
+    {
+        if ( ! is_resource($socket) ) {
+            return false;
+        }
+
+        $data = '';
+        while (substr($data, 3, 1) !== ' ') {
+            if ( ! ( $data = @fgets($socket, 256) ) ) {
+                $this->lastResponseCode = -1;
```

Comment:
Do you actually need this to be class level property? I haven't found this to be "reused", just overwritten. Making it local to the scope will prevent errors or unexpected manipulations from the outside.

Reply:
Yes, is needed property.

Reply:
I think my comment was misleading. The variable is needed, that´s correct. However, instead of using a class property I think it can be a local (that is, declared and used within the method) var.
Using
```php
function Bar {
  $lastResponseCode = null;
}
```
Instead of 
```php
class Foo {
  private $lastResponseCode;
  function Baz {
     $this->lastResponseCode...
  }
}
```